worker/codalabworker/local_run/local_dependency_manager.py

```diff
@@ -0,0 +1,277 @@
+
+class DownloadAbortedException(Exception):
+    """
+    Exception raised by the download if a download is killed before it is complete
+    """
+    def __init__(self, message):
+        super(DownloadAbortedException, self).__init__(message)
+
+class LocalFileSystemDependencyManager(StateTransitioner, BaseDependencyManager):
+    DEPENDENCIES_DIR_NAME = 'dependencies'
+    def __init__(self, state_committer, bundle_service,
+            worker_dir, max_cache_size_bytes, max_serialized_length):
+
+        super(LocalFileSystemDependencyManager, self).__init__()
+        self.add_transition(DependencyStage.DOWNLOADING, self._transition_from_DOWNLOADING)
+        self.add_transition(DependencyStage.READY, self._transition_from_READY)
+        self.add_transition(DependencyStage.FAILED, self._transition_from_FAILED)
+
+        self._state_committer = state_committer
+        self._bundle_service = bundle_service
+        self._max_cache_size_bytes = max_cache_size_bytes
+        self._max_serialized_length = max_serialized_length or float('inf')
+        self.dependencies_dir = os.path.join(worker_dir, LocalFileSystemDependencyManager.DEPENDENCIES_DIR_NAME)
+        if not os.path.exists(self.dependencies_dir):
+            logger.info('{} doesn\'t exist, creating.'.format(self.dependencies_dir))
+            os.makedirs(self.dependencies_dir, 0770)
+
+        self._lock = threading.RLock()
+
+        self._paths = set()
+        self._dependencies = dict()
+        self._downloading = dict()
+        self._load_state()
+
+        self._stop = False
+        self._main_thread = None
+
+    def _save_state(self):
+        with self._lock:
+            self._state_committer.commit(self._dependencies)
+
+    def _load_state(self):
+        with self._lock:
+            self._dependencies = self._state_committer.load()
+            logger.info('{} dependencies in cache.'.format(len(self._dependencies)))
+
+    def start(self):
+        def loop(self):
+            while not self._stop:
+                try:
+                    self._process_dependencies()
+                    self._save_state()
+                    self._cleanup()
+                    self._save_state()
+                except Exception:
+                    traceback.print_exc()
+                # TODO: decide if we need to sleep here
+        self._main_thread = threading.Thread(target=loop, args=[self])
+        self._main_thread.start()
+
+    def stop(self):
+        self._stop = True
+        self._main_thread.join()
+
+    def _process_dependencies(self):
+        with self._lock:
+            for entry, state in self._dependencies.items():
+                self._dependencies[entry] = self.transition(state)
+
+    def _cleanup(self):
+        while True:
+            with self._lock:
+                bytes_used = sum(dep.size_bytes for dep in self._dependencies.values())
+                serialized_dependencies = {'{}+{}'.format(*k): v for k, v in self._dependencies.items()}
```

Comment:
why don't we serialize dependencies directly instead of having `format` as a string?